iocore/net/LuaSNIConfig.cc

```diff
@@ -0,0 +1,125 @@
+/** @file
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#include "LuaSNIConfig.h"
+#include <cstring>
+#include "ts/Diags.h"
+#include "P_SNIActionPerformer.h"
+
+TsConfigDescriptor LuaSNIConfig::desc = {TsConfigDescriptor::Type::ARRAY, "Array", "Item vector", "Vector"};
+TsConfigArrayDescriptor LuaSNIConfig::DESCRIPTOR(LuaSNIConfig::desc);
+TsConfigDescriptor LuaSNIConfig::Item::FQDN_DESCRIPTOR = {TsConfigDescriptor::Type::STRING, "String", TS_fqdn,
+                                                          "Fully Qualified Domain Name"};
+TsConfigDescriptor LuaSNIConfig::Item::DISABLE_h2_DESCRIPTOR = {TsConfigDescriptor::Type::BOOL, "Boolean", TS_disable_H2,
+                                                                "Disable H2"};
+TsConfigEnumDescriptor LuaSNIConfig::Item::LEVEL_DESCRIPTOR = {TsConfigDescriptor::Type::ENUM,
+                                                               "enum",
+                                                               "Level",
+                                                               "Level for client verification",
+                                                               {{"NONE", 0}, {"MODERATE", 1}, {"STRICT", 2}}};
+TsConfigDescriptor LuaSNIConfig::Item::TUNNEL_DEST_DESCRIPTOR = {TsConfigDescriptor::Type::STRING, "String", TS_tunnel_route,
+                                                                 "tunnel route destination"};
+TsConfigDescriptor LuaSNIConfig::Item::CLIENT_CERT_DESCRIPTOR = {TsConfigDescriptor::Type::STRING, "String", TS_client_cert,
+                                                                 "Client certificate to present to the next hop server"};
+TsConfigDescriptor LuaSNIConfig::Item::VERIFY_NEXT_SERVER_DESCRIPTOR = {TsConfigDescriptor::Type::INT, "Int",
+                                                                        TS_verify_origin_server, "Next hop verification level"};
+
+ts::Errata
+LuaSNIConfig::loader(lua_State *L)
```

Comment:
[BindingInstance](https://github.com/apache/trafficserver/blob/master/lib/bindings/bindings.h) is the canonical way to make Lua APIs. Please use that instead.

Reply:
This is based on the TsLuaConfig work which I did a brief presentation on two summits back, and a longer at the most recent summit. I've discussed it at length with Kit as well and he's OK with this approach. This is really a prototype of the work, to get this PR running.

Reply:
I really don't see why we need another way to bind Lua APIs.

Reply:
https://cwiki.apache.org/confluence/display/TS/Presentations+-+2017?preview=/70255385/74684802/TsLuaConfig-Fall-2017.pdf

Reply:
@jpeach - I think the schema based approach is better and it's reasonable to try it while we are still getting set up for using Lua configs.  There's an example of using LuaBindings, I want to try an example using TsLuaConfig.